### PR TITLE
Cloudprovider hetzner

### PR DESCRIPTION
**Pull request by Fgruntjes**
So I made a very basic implementation of a cloud provider for Hetzner Cloud. Now I am not sure if I did everything correctly so please let me know where you see any issues / improvements.

Once you are satisfied with the implementation I will add test cases to finalize.

### CODE REVIEW

**Comment by Fgruntjes** [comment]
@k8s-ci-robot verify sign request

**Review by mwielgus — CHANGES_REQUESTED** [review]
We don't accept any new dependencies in `vendor`. Please contain your dependencies inside `cloudprovider/hetzner`. See Alibaba cloudprovider for reference. https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/alicloud.

**Comment by Fgruntjes** [comment]
> We don't accept any new dependencies in `vendor`. Please contain your dependencies inside `cloudprovider/hetzner`. See Alibaba cloudprovider for reference. https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/alicloud.

Thank you, solved that issue.

**Comment by MaciekPytel** [comment]
Some high-level comments:

* Unfortunately we can't accept any contribution until the CLA tests passes. It seems you may have already signed it?
* We ask that any new provider has an OWNERS file with maintainers of that provider. With so many existing providers we're no longer able to maintain them.
* It seems that hcloud-go is an inlined external client? If so feel free to exclude it from linter checks, etc - you can use PR #3619 as an example.

/check-cla

**Comment by Fgruntjes** [comment]
@MaciekPytel Thank you for the suggestions

Yes, I have signed it but somehow it is not recognizing I now updated our organization with the email domain and set it to public. Let's see if that's enough.

Will add the OWNERS and linter exclude.

/check-cla

**Comment by Jeffwan** [comment]
I feel like this is not a scalable way to manage all cloud providers in-tree now. 
Do we want to onboard user on GRPC solution in the future? https://github.com/kubernetes/autoscaler/pull/3140

**Comment by alexanderkjeldaas** [comment]
@Jeffwan shouldn't the proposal in #3140 be merged, maybe with an example implementation, before blocking new  implementations?

**Comment by Jeffwan** [comment]
@alexanderkjeldaas Yeah, this PR should not be blocked for sure. Seems for the past half year, 2-3 cloud providers are added in the repo, I am thinking sig-autoscaler should prioritize pluggable solutions.

**Comment by niklaskorz** [comment]
I have written an alternative Hetzner autoscaler implementation we have been using successfully since April. Glad to see someone went through to open a pull request, I will have a look at your code and see if I can come up with any useful improvements before this is merged. :) @Fgruntjes

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go
Comment (niklaskorz) [inline cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go:104]: I am not exactly sure what `Pricing()` is used for in autoscaler, but the Hetzner API would make an implementation of this possible: https://docs.hetzner.cloud/#server-types

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go
Comment (niklaskorz) [inline cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go:304]: We could consider including the cluster name in the node name as well, which can be specified through the `--cluster-name` parameter.
On the other hand, it could be considered "best practice" to have only one cluster oer Hetzner Cloud project.
Just food for thought, I don't think it's a necessary change.

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go
Comment (niklaskorz) [inline cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go:355]: Same as above: could include a label for the cluster name as well, but that depends on whether a Hetzner Cloud project should have more than one cluster anyway, or if clusters running in the same project can rely on choosing distinct node group names.
It might just overcomplicate the implementation though, so I'd be fine with leaving it the way it is.

**Review by niklaskorz — CHANGES_REQUESTED** [review]
Really well done! Especially nice to see a fully working node group implementation, my own implementation was limitted to a single node group. I added a few comments, I hope they are helpful.

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go
Comment (niklaskorz) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go:104]: On the other hand, this is not even implemented by Azure or AWS, so pretty save to leave it out...

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go
Comment (WebSpider) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go:304]: Using one cluster per project has some downsides, given that authentication has to be setup manually per project: There are no (documented) API calls to setup authentication tokens for a new project.

File: cluster-autoscaler/cloudprovider/hetzner/hcloud-go/hcloud/action.go
Comment (Bessonov) [inline cluster-autoscaler/cloudprovider/hetzner/hcloud-go/hcloud/action.go:2]: It seems like that `hcloud-go` is a copy from https://github.com/hetznercloud/hcloud-go . It's licensed under MIT with "License and copyright notice". I'm not sure that hcloud-go can be directly copy&pasted with a license change.

**Comment by Bessonov** [comment]
@mwielgus 

> We don't accept any new dependencies in vendor. Please contain your dependencies inside cloudprovider/hetzner. See Alibaba cloudprovider for reference. https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/alicloud.

Is there any reason for that? How it's intended to be updated?

File: cluster-autoscaler/cloudprovider/builder/builder_hetzner.go
Comment (Bessonov) [inline cluster-autoscaler/cloudprovider/builder/builder_hetzner.go:1]:
```suggestion
// +build hetzner
```

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go
Comment (Bessonov) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go:104]: See https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/pricing.md

Comment (niklaskorz) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go:104]: Oh, that's pretty awesome actually, and would be a perfect fit here as the prices are indeed available through the API.

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go
Comment (Fgruntjes) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go:304]: What @WebSpider said, we can not create a new project and api key so to ensure we need as little manual intervention as possible I decided to go for node labels.

File: cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go
Comment (Fgruntjes) [reply cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go:104]: Yes it would definitely be possible, we did not need this for our use case, so it was skipped. Feel free to implement, create a PR on our repo and I am happy to include.

**Comment by mwielgus** [comment]
You are responsible for updating the dependencies once/if they are needed. We don't want to onboard any new libraries as they may have conflicting dependencies with other libraries. Imagine that both cloud provider A and B SDK requires library X. However A is only compatible with 1.0. B SDK has been migrated to 2.0. But Cluster Autoscaler still is on old B SDK with X in version 1.0.   Now you want to update B due to security issues. How would you do that? There is no easy solution. You cannot kick out cloud provider A. You cannot have 2 libraries in the same binary. You have to upgrade B. 
To avoid such problems we don't accept any dependencies of cloud providers. Everything needs to be contained in their directory. If this doesn't work for you, CA is open source and can be forked if needed :).

**Comment by mwielgus** [comment]
Please resolve conflicts and squash commits to max 3 and we are good to go.

**Comment by LKaemmerling** [comment]
Hello guys,

@Fgruntjes asked us (Hetzner Cloud Team) to take over this MR. 

We (my colleague @fhofherr and I) want to take this over and further support it and also update everything in the future. So the plan was that we (@fhofherr and i (@LKaemmerling)) are set within the code owners file.

@Fgruntjes thank you again for the time you invested into this! You did a great job and we are really happy that we can take it over and make it officially supported by our team. 

Thank you!

**Comment by Fgruntjes** [comment]
@LKaemmerling thnx for your comment. Ill close this PR so you guys can create a new one to take over.

**Pull request closed without merging**
